src/high-level/polynomial-solver.lisp

```diff
@@ -0,0 +1,112 @@
+;;;; src/polynomial-solver.lisp
+;;;;
+;;;; Author: Juan M. Bello-Rivas
+
+(in-package #:magicl)
+
+;;; This file implements a simple univariate polynomial solver that relies on
+;;; the eigendecomposition of the companion matrix of the given polynomial to
+;;; obtain all of its roots.
+
+;;; A root refinement mechanism via Newton's iteration is included. Note that
+;;; polynomials (a) whose leading coefficient is much smaller than the
+;;; remaining coefficients or (b) having roots with multiplicity higher than
+;;; one will be harder to solve accurately.
+
+;;; A univariate polynomial is represented as a one-dimensional array of
+;;; complex coefficients.
+
+(deftype polynomial-coefficients () '(simple-array (complex double-float) (*)))
+
+(defstruct (polynomial (:constructor %make-polynomial))
+  (coefficients nil :type polynomial-coefficients))
+
+(defun make-polynomial (&rest args)
```

Comment:
Our proposed use-case is for polynomials of degree `2^n - 1`. I don't think this should be ` &rest`, lest we blow the stack.

Reply:
I agree but I think it is OK to have `make-polynomial` be an easy-to-use constructor for tinkering in the REPL while letting the low-level constructor `%make-polynomial` handle the case where efficiency is needed.

Reply:
It’s just the name that spooks me, I guess. The `%` looks more serious to me than “my argument signature is un-flattened”.

Reply:
I'm open to suggestions regarding nomenclature!